docs/tutorial/typesfuns.rst

```diff
@@ -661,6 +661,93 @@ without any explicit use of ``Force`` or ``Delay``:
     ifThenElse True  t e = t;
     ifThenElse False t e = e;
 
+Codata Types
+============
+
+Codata types are like regular data types, except that they allow for us to
+define infinite data structures.  More precisely, for a type ``T``, each of its
+constructor arguments of type ``T`` are transformed into a coinductive parameter
+``Inf T``. This makes each of the ``T`` arguments lazy, and allows infinite data
```

Comment:
A)
The explanation of why codata can not allow for infinite mutual recursive
definitions doesn't look good.
Codata is supposed to be sugar for Inf, thus it doesn't make sense for it
to not work.

In the definition of Mutual it specifies that dependent type reduction
happen after type declarations are added. Does codata need to be reduced to
the desugared version? If so, it is very bad from a pedagogical point to
have to understand the implementation of codata to make sense of the rules.

B) There are two more examples that are needed:

1) An example of a type that has one constructor with finite recursion and
another with infinite corecursion.
https://github.com/idris-lang/Idris-dev/blob/da914af83e36d726ba352fb89f5c15689b675540/test/totality004/totality004.idr

2) One also needs to expand the totality definition to add functions that
have codata in their corecursions.

My opinion is that the concept of totality should be removed from the
Theorem proving section and become an independent one.
This would also fix #2983.

On Thu, Feb 25, 2016 at 10:19 PM, Leif Warner <notifications@github.com>
wrote:

> In docs/tutorial/typesfuns.rst
> <https://github.com/idris-lang/Idris-dev/pull/2988#discussion_r54157633>:
>
> > @@ -661,6 +661,93 @@ without any explicit use of ``Force`` or ``Delay``:
> >      ifThenElse True  t e = t;
> >      ifThenElse False t e = e;
> >
> > +Codata Types
> > +============
> > +
> > +Codata types are like regular data types, except that they allow for us to
> > +define infinite data structures.  More precisely, for a type ``T``, each of its
> > +constructor arguments of type ``T`` are transformed into a coinductive parameter
> > +``Inf T``. This makes each of the ``T`` arguments lazy, and allows infinite data
>
> I think I like the precise language. We're already talking about codata
> here; I think they can get the gist from the context on this that the
> recursive position is treated a little differently; worst case they can go
> and look up coinduction.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idris-lang/Idris-dev/pull/2988/files#r54157633>.
>